src/katsdpmodels/fetch/__init__.py

```diff
@@ -263,3 +264,56 @@ def _close(self) -> None:
             self._model_cache.clear()
         else:
             self._model_cache = {}     # Allow garbage collection of the old cache
+
+
+class TelescopeStateRequest(Generic[_TS]):
+    """A request for a key from a telescope state."""
+
+    def __init__(self, telstate: _TS, key: str) -> None:
+        self.telstate = telstate
+        self.key = key
+
+
+_TSGenerator = Generator[TelescopeStateRequest[_TS], object, str]
+
+
+class TelescopeStateFetcherBase(Generic[_TS]):
+    """Base class for fetching models from telescope state.
+
+    This is a low-level class not intended for direct usage. It handles
+    the mechanics of deciding which telescope state keys are needed and
+    assembling them into an URL. The details of how to get data out of
+    telescope state (either synchronously or asynchronously) and passing
+    that to a model class is left to the subclass.
+
+    This class is *not* thread-safe.
+    """
+
+    def __init__(self, telstate: _TS) -> None:
+        self.telstate = telstate
+        # Cache of sdp_model_base_url
+        self._base_url: Optional[str] = None
+
+    def _get_str(self, telstate: _TS, key: str) -> _TSGenerator:
+        """Get a string-valued key from the telescope state.
+
+        It yields the key to the caller, which should either respond
+        with the value or throw in an exception.
+        """
+        try:
+            value = yield TelescopeStateRequest(telstate, key)
+        except Exception as exc:
+            raise models.TelescopeStateError(f'could not get {key}: {exc}') from exc
+        if not isinstance(value, str):
+            type_ = type(value)
+            raise models.TelescopeStateError(f'{key} should be a str, not {type_}')
+        else:
+            return value
+
+    def _get_url(self, key: str, *, telstate: Optional[_TS] = None) -> _TSGenerator:
+        if self._base_url is None:
+            self._base_url = yield from self._get_str(self.telstate, 'sdp_model_base_url')
+        if telstate is None:
```

Comment:
Yes, it's intentional - although typically they'll be different views rather than different backends. In katsdpingest there is a global telstate view (for sdp_model_base_url) and there is a "CBF" view (that views only the relevant CBF streams, and which is appropriate for the band mask key).